# IPL Match Result Prediction 

Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Reading File

In [2]:
df = pd.read_csv('SigTuple_train.csv')
df.head()

Game ID                        Team 1                       Team 2  \
0        1   Koramangala Traffic Jammers    Whitefield Water Loggers    
1        2  Electronic City Power Savers        Silkboard Slow Movers   
2        3       Indranagar Pub Watchers       Sarjapur Water Tankers   
3        4      Bellandur Froth Fighters  Koramangala Traffic Jammers   
4        5           Marathalli Chokers     Whitefield Water Loggers    

          City  DayOfWeek  DateOfGame TimeOfGame  AvgWindSpeed  AvgHumidity  \
0   Whitefield          1  01-01-2012   20:00:00             6         0.49   
1    Silkboard          2  01-02-2012   17:00:00             7         0.44   
2     Sarjapur          3  01-03-2012   20:30:00            11         0.23   
3  Koramangala          4  01-04-2012   16:00:00             6         0.61   
4   Marathalli          5  01-05-2012   20:00:00             6         0.56   

   Inn 1 Team 1 NOP R>25,SR>125             ...               \
0                             1             ...                
1                             3             ...                
2                             2             ...                
3                             0             ...                
4                             3             ...                

   Inn 2 Team 2 NOP R<25, SR>125  Inn 2 Team 2 Total 4s  \
0                              0                      3   
1                              2                     18   
2                              1                     18   
3                              0                      5   
4                              3                     15   

   Inn 2 Team 2 Total 6s  Inn 2 Team 2 Max Strike Rate_ALLBatsmen  \
0                      3                                   120.00   
1                      9                                   215.15   
2                      1                                   300.00   
3                      4                                   100.00   
4                      6                                   205.26   

   Inn 2 Team 1 NoP fast bowlers  Inn 2 Team 1 NoP Spinners  \
0                              5                          0   
1                              4                          1   
2                              3                          3   
3                              4                          2   
4                              4                          2   

   Inn 2 Team 1 wickets taken_catches_runout  \
0                                          6   
1                                          4   
2                                          0   
3                                          3   
4                                          4   

   Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping  \
0                                                  4                  
1                                                  0                  
2                                                  1                  
3                                                  2                  
4                                                  1                  

   Inn 2 Team 1 Extras conceded in_wides_No Balls  Winner (team 1=1, team 2=0)  
0                                              11                            1  
1                                               5                            1  
2                                              10                            0  
3                                              16                            0  
4                                               5                            0  

[5 rows x 30 columns]

In [4]:
# What do the features mean?
features_meaning = pd.read_excel('data_dictionary.xlsx')
features_meaning

Variable Name  \
0                                              Team 1   
1                                              Team 2   
2                                       City of match   
3                                                 Day   
4                                       Date of Match   
5                                       Time of Match   
6                                      Avg Wind Speed   
7                                        Avg Humidity   
8                        Inn 1 Team 1 NOP R>25,SR>125   
9                       Inn 1 Team 1 NOP R<25, SR>125   
10                              Inn 1 Team 1 Total 4s   
11                              Inn 1 Team 1 Total 6s   
12            Inn 1 Team 1 Max Strike Rate_ALLBatsmen   
13                      Inn 1 Team 2 NoP fast bowlers   
14                          Inn 1 Team 2 NoP Spinners   
15          Inn 1 Team 2 wickets taken_catches_runout   
16  Inn1 Team 2 wickets taken_ bowled _lbw_caught ...   
17     Inn 1 Team 2 Extras conceded in_wides_No Balls   
18                       Inn 2 Team 2 NOP R>25,SR>125   
19                      Inn 2 Team 2 NOP R<25, SR>125   
20                              Inn 2 Team 2 Total 4s   
21                              Inn 2 Team 2 Total 6s   
22            Inn 2 Team 2 Max Strike Rate_ALLBatsmen   
23                      Inn 2 Team 1 NoP fast bowlers   
24                          Inn 2 Team 1 NoP Spinners   
25          Inn 2 Team 1 wickets taken_catches_runout   
26  Inn 2 Team 1 wickets taken_ bowled _lbw_caught...   
27     Inn 2 Team 1 Extras conceded in_wides_No Balls   
28                      Team won (team 1=1, team 2=0)   

                                          Description  \
0               Team batting first and bowling second   
1               Team batting second and bowling first   
2                      City where the match is played   
3            Day of the week when the match is played   
4           Date of the year when the match is played   
5            Time of the day when the match is played   
6   Average speed of the wind on the day when the ...   
7   Average humidity on the day when the match is ...   
8   NoP(Number of players) in Team 1 that scored m...   
9   NoP(Number of players) in Team 1 that scored l...   
10  Total Number of 4s hit by Team 1 in the first ...   
11  Total Number of 6s hit by Team 1 in the first ...   
12  Maximum strike rate achieved including all bat...   
13  NoP(Number of players) in Team 2 who are fast ...   
14  NoP(Number of players) in Team 2 who are spinn...   
15  Number of wickets taken by Team 2 in the first...   
16  Number of wickets taken by Team 2 in the first...   
17  Total Number Extra runs conceded by Team 2 in ...   
18  NoP(Number of players) in Team 2 that scored m...   
19  NoP(Number of players) in Team 2 that scored l...   
20  Total Number of 4s hit by Team 2 in the second...   
21  Total Number of 6s hit by Team 2 in the second...   
22  Maximum strike rate achieved including all bat...   
23  NoP(Number of players) in Team 1 who are fast ...   
24  NoP(Number of players) in Team 1 who are spinn...   
25  Number of wickets taken by Team 1 in the secon...   
26  Number of wickets taken by Team 1 in the secon...   
27  Total Number Extra runs conceded by Team 1 in ...   
28  If Team batting first i.e. Team1 won the match...   

                                   Cricketing Jargons  
0                                                 NaN  
1   Bowled\na mode of a batsman's dismissal. Occur...  
2   Catch\nto dismiss a batsman by a fielder catch...  
3   Run out\ndismissal by a member of the fielding...  
4   Leg before wicket (LBW)\na way of dismissing t...  
5   Stumping: It requires co-operation between a b...  
6   Strike rate\n percentage equal to the number o...  
7   Innings\none player's or one team's turn to ba...  
8   Spin bowling\na style of bowling in which a sp...  
9                                                 NaN  
10

Exploring Data

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252 entries, 0 to 251
Data columns (total 30 columns):
Game ID                                                             252 non-null int64
Team 1                                                              252 non-null object
Team 2                                                              252 non-null object
City                                                                252 non-null object
DayOfWeek                                                           252 non-null int64
DateOfGame                                                          252 non-null object
TimeOfGame                                                          252 non-null int64
AvgWindSpeed                                                        252 non-null int64
AvgHumidity                                                         252 non-null float64
Inn 1 Team 1 NOP R>25,SR>125                                        252 non-null int64
Inn 1 Team 1 NOP R<

In [3]:
df['TimeOfGame'].value_counts()

20:00:00    133
16:00:00     55
16:30:00     25
12:30:00     22
16:45:00     10
20:30:00      2
17:10:00      1
17:00:00      1
20:15:00      1
15:00:00      1
18:50:00      1
Name: TimeOfGame, dtype: int64

Feature Engineering

We help model classify based on match played during 1) noon, 2) afternoon/evening or 3) late evening/night

In [4]:
df['TimeOfGame'] = df['TimeOfGame'].replace('12:30:00', 0)
df['TimeOfGame'] = df['TimeOfGame'].replace('15:00:00', 1)
df['TimeOfGame'] = df['TimeOfGame'].replace('16:00:00', 1)
df['TimeOfGame'] = df['TimeOfGame'].replace('16:30:00', 1)
df['TimeOfGame'] = df['TimeOfGame'].replace('16:45:00', 1)
df['TimeOfGame'] = df['TimeOfGame'].replace('17:00:00', 1)
df['TimeOfGame'] = df['TimeOfGame'].replace('17:10:00', 1)
df['TimeOfGame'] = df['TimeOfGame'].replace('18:50:00', 2)
df['TimeOfGame'] = df['TimeOfGame'].replace('20:00:00', 2)
df['TimeOfGame'] = df['TimeOfGame'].replace('20:15:00', 2)
df['TimeOfGame'] = df['TimeOfGame'].replace('20:30:00', 2)
df['TimeOfGame'].value_counts()

2    137
1     93
0     22
Name: TimeOfGame, dtype: int64

In [5]:
df['Team 1'].value_counts(), df['Team 2'].value_counts()

(Electronic City Power Savers    39
 Marathalli Chokers              33
 Koramangala Traffic Jammers     33
 Bellandur Froth Fighters        32
 Silkboard Slow Movers           28
 Indranagar Pub Watchers         26
 Sarjapur Water Tankers          26
 Whitefield Water Loggers        22
 Airport Flyers                   7
 Forum Fans                       6
 Name: Team 1, dtype: int64, Whitefield Water Loggers        41
 Sarjapur Water Tankers          32
 Indranagar Pub Watchers         32
 Silkboard Slow Movers           29
 Bellandur Froth Fighters        28
 Marathalli Chokers              27
 Koramangala Traffic Jammers     24
 Electronic City Power Savers    24
 Forum Fans                       8
 Airport Flyers                   7
 Name: Team 2, dtype: int64)

In [7]:
df.groupby(['Winner (team 1=1, team 2=0)']).mean()

Game ID  DayOfWeek  TimeOfGame  AvgWindSpeed  \
Winner (team 1=1, team 2=0)                                                    
0                            123.645390   4.141844    1.432624      7.418440   
1                            130.126126   3.846847    1.486486      7.468468   

                             AvgHumidity  Inn 1 Team 1 NOP R>25,SR>125  \
Winner (team 1=1, team 2=0)                                              
0                               0.626596                      1.078014   
1                               0.623423                      1.882883   

                             Inn 1 Team 1 NOP R<25, SR>125  \
Winner (team 1=1, team 2=0)                                  
0                                                 1.375887   
1                                                 1.729730   

                             Inn 1 Team 1 Total 4s  Inn 1 Team 1 Total 6s  \
Winner (team 1=1, team 2=0)                                                 
0                                        12.404255               3.801418   
1                                        15.405405               6.252252   

                             Inn 1 Team 1 Max Strike Rate_ALLBatsmen  \
Winner (team 1=1, team 2=0)                                            
0                                                         184.821135   
1                                                         215.902252   

                                                  ...                        \
Winner (team 1=1, team 2=0)                       ...                         
0                                                 ...                         
1                                                 ...                         

                             Inn 2 Team 2 NOP R>25,SR>125  \
Winner (team 1=1, team 2=0)                                 
0                                                1.248227   
1                                                1.144144   

                             Inn 2 Team 2 NOP R<25, SR>125  \
Winner (team 1=1, team 2=0)                                  
0                                                 0.964539   
1                                                 1.612613   

                             Inn 2 Team 2 Total 4s  Inn 2 Team 2 Total 6s  \
Winner (team 1=1, team 2=0)                                                 
0                                        13.248227               4.617021   
1                                        11.909910               4.396396   

                             Inn 2 Team 2 Max Strike Rate_ALLBatsmen  \
Winner (team 1=1, team 2=0)                                            
0                                                         189.977730   
1                                                         186.643874   

                             Inn 2 Team 1 NoP fast bowlers  \
Winner (team 1=1, team 2=0)                                  
0                                                 3.631206   
1                                                 3.729730   

                             Inn 2 Team 1 NoP Spinners  \
Winner (team 1=1, team 2=0)                              
0                                             2.141844   
1                                             2.153153   

                             Inn 2 Team 1 wickets taken_catches_runout  \
Winner (team 1=1, team 2=0)                                              
0                                                             1.978723   
1                                                             5.000000   

                             Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping  \
Winner (team 1=1, team 2=0)                                                                     
0                                                                     1.319149                  
1                                                                     3.072072          

We classify the teams which bat first as 1 and those bowled first as -1

In [9]:
condition1 = [df['Team 1']=='Electronic City Power Savers', df['Team 2']=='Electronic City Power Savers']
choice1 = ['1', '-1']
df['Electronic City Power Savers_new'] = np.select(condition1, choice1, default = 0)
condition2 = [df['Team 1']=='Marathalli Chokers', df['Team 2']=='Marathalli Chokers']
choice2 = ['1', '-1']
df['Marathalli Chokers_new'] = np.select(condition2, choice2, default = 0)
condition3 = [df['Team 1']=='Koramangala Traffic Jammers', df['Team 2']=='Koramangala Traffic Jammers']
choice3 = ['1', '-1']
df['Koramangala Traffic Jammers_new'] = np.select(condition3, choice3, default = 0)
condition4 = [df['Team 1']=='Bellandur Froth Fighters', df['Team 2']=='Bellandur Froth Fighters']
choice4 = ['1', '-1']
df['Bellandur Froth Fighters_new'] = np.select(condition4, choice4, default = 0)
condition5 = [df['Team 1']=='Silkboard Slow Movers', df['Team 2']=='Silkboard Slow Movers']
choice5 = ['1', '-1']
df['Silkboard Slow Movers_new'] = np.select(condition5, choice5, default = 0)
condition6 = [df['Team 1']=='Indranagar Pub Watchers', df['Team 2']=='Indranagar Pub Watchers']
choice6 = ['1', '-1']
df['Indranagar Pub Watchers_new'] = np.select(condition6, choice6, default = 0)
condition7 = [df['Team 1']=='Forum Fans', df['Team 2']=='Forum Fans']
choice7 = ['1', '-1']
df['Forum Fans_new'] = np.select(condition7, choice7, default = 0)
condition8 = [df['Team 1']=='Airport Flyers', df['Team 2']=='Airport Flyers']
choice8 = ['1', '-1']
df['Airport Flyers_new'] = np.select(condition8, choice8, default = 0)
condition9 = [df['Team 1']=='Whitefield Water Loggers', df['Team 2']=='Whitefield Water Loggers']
choice9 = ['1', '-1']
df['Whitefield Water Loggers_new'] = np.select(condition9, choice9, default = 0)
condition10 = [df['Team 1']=='Sarjapur Water Tankers', df['Team 2']=='Sarjapur Water Tankers']
choice10 = ['1', '-1']
df['Sarjapur Water Tankers_new'] = np.select(condition10, choice10, default = 0)
df.head()

Game ID                        Team 1                       Team 2  \
0        1   Koramangala Traffic Jammers    Whitefield Water Loggers    
1        2  Electronic City Power Savers        Silkboard Slow Movers   
2        3       Indranagar Pub Watchers       Sarjapur Water Tankers   
3        4      Bellandur Froth Fighters  Koramangala Traffic Jammers   
4        5           Marathalli Chokers     Whitefield Water Loggers    

          City  DayOfWeek  DateOfGame  TimeOfGame  AvgWindSpeed  AvgHumidity  \
0   Whitefield          1  01-01-2012           2             6         0.49   
1    Silkboard          2  01-02-2012           1             7         0.44   
2     Sarjapur          3  01-03-2012           2            11         0.23   
3  Koramangala          4  01-04-2012           1             6         0.61   
4   Marathalli          5  01-05-2012           2             6         0.56   

   Inn 1 Team 1 NOP R>25,SR>125             ...              \
0                             1             ...               
1                             3             ...               
2                             2             ...               
3                             0             ...               
4                             3             ...               

   Electronic City Power Savers_new  Marathalli Chokers_new  \
0                                 0                       0   
1                                 1                       0   
2                                 0                       0   
3                                 0                       0   
4                                 0                       0   

   Koramangala Traffic Jammers_new  Bellandur Froth Fighters_new  \
0                                1                             0   
1                                0                             0   
2                                0                             0   
3                               -1                             1   
4                                0                             0   

   Silkboard Slow Movers_new  Indranagar Pub Watchers_new  Forum Fans_new  \
0                          0                            0               0   
1                         -1                            0               0   
2                          0                            1               0   
3                          0                            0               0   
4                          0                            0               0   

   Airport Flyers_new  Whitefield Water Loggers_new  \
0                   0                             0   
1                   0                             0   
2                   0                             0   
3                   0                             0   
4                   0                             0   

   Sarjapur Water Tankers_new  
0                           0  
1                           0  
2                          -1  
3                           0  
4                           0  

[5 rows x 40 columns]

We one hot encode Cities

In [10]:
df1 = pd.get_dummies(df['City'], dummy_na=True)
df = pd.concat([df, df1], axis=1)
df.head()

Game ID                        Team 1                       Team 2  \
0        1   Koramangala Traffic Jammers    Whitefield Water Loggers    
1        2  Electronic City Power Savers        Silkboard Slow Movers   
2        3       Indranagar Pub Watchers       Sarjapur Water Tankers   
3        4      Bellandur Froth Fighters  Koramangala Traffic Jammers   
4        5           Marathalli Chokers     Whitefield Water Loggers    

          City  DayOfWeek  DateOfGame  TimeOfGame  AvgWindSpeed  AvgHumidity  \
0   Whitefield          1  01-01-2012           2             6         0.49   
1    Silkboard          2  01-02-2012           1             7         0.44   
2     Sarjapur          3  01-03-2012           2            11         0.23   
3  Koramangala          4  01-04-2012           1             6         0.61   
4   Marathalli          5  01-05-2012           2             6         0.56   

   Inn 1 Team 1 NOP R>25,SR>125 ...   Sarjapur Water Tankers_new  Bellandur  \
0                             1 ...                            0          0   
1                             3 ...                            0          0   
2                             2 ...                           -1          0   
3                             0 ...                            0          0   
4                             3 ...                            0          0   

   Electronic City  Indranagar  Koramangala  Marathalli  Sarjapur  Silkboard  \
0                0           0            0           0         0          0   
1                0           0            0           0         0          1   
2                0           0            0           0         1          0   
3                0           0            1           0         0          0   
4                0           0            0           1         0          0   

   Whitefield  NaN  
0           1    0  
1           0    0  
2           0    0  
3           0    0  
4           0    0  

[5 rows x 49 columns]

Extract features out of Date

In [11]:
df['yearOfGame'] = pd.to_datetime(df['DateOfGame']).dt.year
df['monthOfGame'] = pd.to_datetime(df['DateOfGame']).dt.month

In [12]:
df = df.drop(['Game ID'], axis=1)
df = df.drop(['Team 1'], axis=1)
df = df.drop(['Team 2'], axis=1)
df = df.drop(['City'], axis=1)
df = df.drop(['DateOfGame'], axis=1)

In our conversion as -1 and 1 for teams which bowled first and batted first respectively, the returned data type is uint. We convert this to int.

In [13]:
df['Electronic City Power Savers_new'] = df['Electronic City Power Savers_new'].apply(lambda x: int(x))
df['Marathalli Chokers_new'] = df['Marathalli Chokers_new'].apply(lambda x: int(x))
df['Koramangala Traffic Jammers_new'] = df['Koramangala Traffic Jammers_new'].apply(lambda x: int(x))
df['Bellandur Froth Fighters_new'] = df['Bellandur Froth Fighters_new'].apply(lambda x: int(x))
df['Silkboard Slow Movers_new'] = df['Silkboard Slow Movers_new'].apply(lambda x: int(x))
df['Indranagar Pub Watchers_new'] = df['Indranagar Pub Watchers_new'].apply(lambda x: int(x))
df['Forum Fans_new'] = df['Forum Fans_new'].apply(lambda x: int(x))
df['Airport Flyers_new'] = df['Airport Flyers_new'].apply(lambda x: int(x))
df['Whitefield Water Loggers_new'] = df['Whitefield Water Loggers_new'].apply(lambda x: int(x))
df['Sarjapur Water Tankers_new'] = df['Sarjapur Water Tankers_new'].apply(lambda x: int(x))

In [14]:
df.head()

DayOfWeek  TimeOfGame  AvgWindSpeed  AvgHumidity  \
0          1           2             6         0.49   
1          2           1             7         0.44   
2          3           2            11         0.23   
3          4           1             6         0.61   
4          5           2             6         0.56   

   Inn 1 Team 1 NOP R>25,SR>125  Inn 1 Team 1 NOP R<25, SR>125  \
0                             1                              1   
1                             3                              2   
2                             2                              2   
3                             0                              2   
4                             3                              1   

   Inn 1 Team 1 Total 4s  Inn 1 Team 1 Total 6s  \
0                     15                     14   
1                     20                     16   
2                     13                      3   
3                      6                      6   
4                     18                      5   

   Inn 1 Team 1 Max Strike Rate_ALLBatsmen  Inn 1 Team 2 NoP fast bowlers  \
0                                   216.43                              4   
1                                   246.15                              5   
2                                   200.00                              4   
3                                   140.00                              3   
4                                   233.33                              5   

      ...       Electronic City  Indranagar  Koramangala  Marathalli  \
0     ...                     0           0            0           0   
1     ...                     0           0            0           0   
2     ...                     0           0            0           0   
3     ...                     0           0            1           0   
4     ...                     0           0            0           1   

   Sarjapur  Silkboard  Whitefield  NaN  yearOfGame  monthOfGame  
0         0          0           1    0        2012            1  
1         0          1           0    0        2012            1  
2         1          0           0    0        2012            1  
3         0          0           0    0        2012            1  
4         0          0           0    0        2012            1  

[5 rows x 46 columns]

In [17]:
cols = list(df.columns.values)
cols

['DayOfWeek',
 'TimeOfGame',
 'AvgWindSpeed',
 'AvgHumidity',
 'Inn 1 Team 1 NOP R>25,SR>125',
 'Inn 1 Team 1 NOP R<25, SR>125',
 'Inn 1 Team 1 Total 4s',
 'Inn 1 Team 1 Total 6s',
 'Inn 1 Team 1 Max Strike Rate_ALLBatsmen',
 'Inn 1 Team 2 NoP fast bowlers',
 'Inn 1 Team 2 NoP Spinners',
 'Inn 1 Team 2 wickets taken_catches_runout',
 'Inn1 Team 2 wickets taken_ bowled _lbw_caught by keeper_stumping',
 'Inn 1 Team 2 Extras conceded in_wides_No Balls',
 'Inn 2 Team 2 NOP R>25,SR>125',
 'Inn 2 Team 2 NOP R<25, SR>125',
 'Inn 2 Team 2 Total 4s',
 'Inn 2 Team 2 Total 6s',
 'Inn 2 Team 2 Max Strike Rate_ALLBatsmen',
 'Inn 2 Team 1 NoP fast bowlers',
 'Inn 2 Team 1 NoP Spinners',
 'Inn 2 Team 1 wickets taken_catches_runout',
 'Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping',
 'Inn 2 Team 1 Extras conceded in_wides_No Balls',
 'Winner (team 1=1, team 2=0)',
 'Electronic City Power Savers_new',
 'Marathalli Chokers_new',
 'Koramangala Traffic Jammers_new',
 'Bellandur Froth Fi

Model accuracy of our tree based models also depends on relative position of our features. I initially got 98% accuracy, but unable to track that file with that position set. This is my re-attempt to match the positions as much as possible with that in previous file. 

In [18]:
df = df[['Bellandur','Electronic City','Indranagar','Koramangala','Marathalli','Sarjapur','Silkboard','Whitefield', 'DayOfWeek','TimeOfGame','yearOfGame',
 'monthOfGame','AvgWindSpeed','AvgHumidity','Inn 1 Team 1 NOP R>25,SR>125','Inn 1 Team 1 NOP R<25, SR>125',
 'Inn 1 Team 1 Total 4s',
 'Inn 1 Team 1 Total 6s',
 'Inn 1 Team 1 Max Strike Rate_ALLBatsmen',
 'Inn 1 Team 2 NoP fast bowlers',
 'Inn 1 Team 2 NoP Spinners',
 'Inn 1 Team 2 wickets taken_catches_runout',
 'Inn1 Team 2 wickets taken_ bowled _lbw_caught by keeper_stumping',
 'Inn 1 Team 2 Extras conceded in_wides_No Balls',
 'Inn 2 Team 2 NOP R>25,SR>125',
 'Inn 2 Team 2 NOP R<25, SR>125',
 'Inn 2 Team 2 Total 4s',
 'Inn 2 Team 2 Total 6s',
 'Inn 2 Team 2 Max Strike Rate_ALLBatsmen',
 'Inn 2 Team 1 NoP fast bowlers',
 'Inn 2 Team 1 NoP Spinners',
 'Inn 2 Team 1 wickets taken_catches_runout',
 'Inn2 Team 1 wickets taken_ bowled _lbw_caught by keeper_stumping',
 'Inn 2 Team 1 Extras conceded in_wides_No Balls','Electronic City Power Savers_new',
 'Marathalli Chokers_new',
 'Koramangala Traffic Jammers_new',
 'Bellandur Froth Fighters_new',
 'Silkboard Slow Movers_new',
 'Indranagar Pub Watchers_new',
 'Forum Fans_new',
 'Airport Flyers_new',
 'Whitefield Water Loggers_new',
 'Sarjapur Water Tankers_new','Winner (team 1=1, team 2=0)']]

*Train-Test Split*

In [19]:
from sklearn.model_selection import train_test_split
y = df['Winner (team 1=1, team 2=0)']

In [20]:
X=df.drop(['Winner (team 1=1, team 2=0)'], axis=1)

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [22]:
from sklearn.metrics import accuracy_score, log_loss, mean_squared_error as mse
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
import lightgbm as lgb
from xgboost import XGBClassifier

*Applying and testing models*

In [24]:
import math
classifiers = [DecisionTreeClassifier(), RandomForestClassifier(), AdaBoostClassifier(), GradientBoostingClassifier(), LogisticRegression(penalty='l2'), LogisticRegression(penalty='l1')]
for clf in classifiers:
    clf.fit(X_train, y_train)
    print(clf)
    y_pred = clf.predict(X_test)
    print('Accuracy:', accuracy_score(y_pred, y_test))
    print('Log_loss:', log_loss(y_pred, y_test))
    print('RMSE:', math.sqrt(mse(y_pred, y_test)))

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')
Accuracy: 0.7647058823529411
Log_loss: 8.126864986735479
RMSE: 0.48507125007266594
RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)
Accuracy: 0.803921568627451
Log_loss: 6.772450202470169
RMSE: 0.44280744277004763
AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,


C:\Users\Harsh\.LSC\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)
Accuracy: 0.9215686274509803
Log_loss: 2.7089863523404616
RMSE: 0.28005601680560194
LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l1', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)
Accuracy: 0.9411764705882353
Log_loss: 2.0317397642553465
RMSE: 0.24253562503633297


In [25]:
clf1 = lgb.LGBMClassifier()
clf1.fit(X_train, y_train)
y_pred1 = clf1.predict(X_test)
print('Accuracy:', accuracy_score(y_pred1, y_test))
print('Log_loss:', log_loss(y_pred1, y_test))
print('RMSE:', math.sqrt(mse(y_pred1, y_test)))

Accuracy: 0.9607843137254902
Log_loss: 1.3544931761702315
RMSE: 0.19802950859533486


In [26]:
from sklearn.neural_network import MLPClassifier
mlp=MLPClassifier()
mlp.fit(X_train, y_train)
y_pred2 = mlp.predict(X_test)
print('Accuracy:', accuracy_score(y_pred2, y_test))
print('RMSE:', math.sqrt(mse(y_pred2, y_test)))

Accuracy: 0.47058823529411764
RMSE: 0.7276068751089989


*Trying with Ensembles*

In [27]:
from sklearn.ensemble import VotingClassifier
estimators=[('LGBM', lgb.LGBMClassifier()), ('LOGReg', LogisticRegression(penalty='l2')), ('AdaBoost', AdaBoostClassifier())]
ensemble1 = VotingClassifier(estimators, voting='hard')
ensemble1.fit(X_train, y_train)
ensemble1.score(X_test, y_test)

C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.9411764705882353

In [28]:
from mlens.ensemble import SuperLearner
ensemble2 = SuperLearner()
ensemble2.add([lgb.LGBMClassifier(), LogisticRegression(penalty='l2'), AdaBoostClassifier(), GradientBoostingClassifier()])
ensemble2.add_meta(lgb.LGBMClassifier())
ensemble2.fit(X_train, y_train)
preds = ensemble2.predict(X_test)
math.sqrt(mse(y_test, preds))
accuracy_score(y_test, preds)

[MLENS] backend: threading
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.9019607843137255

In [29]:
ensemble3 = SuperLearner()
ensemble3.add([lgb.LGBMClassifier(), LogisticRegression(penalty='l2'), lgb.LGBMClassifier(), LogisticRegression(penalty='l2')])
ensemble3.add_meta(lgb.LGBMClassifier())
ensemble3.fit(X_train, y_train)
preds1 = ensemble3.predict(X_test)
math.sqrt(mse(y_test, preds1))
accuracy_score(y_test, preds1)

C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarn

0.9215686274509803

In [31]:
estimators4=[('LGBM', lgb.LGBMClassifier()), ('LOGReg', LogisticRegression(penalty='l1'))]
ensemble4 = VotingClassifier(estimators4, voting='hard')
ensemble4.fit(X_train, y_train)
ensemble4.score(X_test, y_test)

C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.9215686274509803

In [40]:
from mlens.ensemble import SequentialEnsemble
ensemble5 = SequentialEnsemble()
ensemble5.add('blend', [lgb.LGBMClassifier(), LogisticRegression(penalty='l1')])
ensemble5.add('stack', [lgb.LGBMClassifier(), lgb.LGBMClassifier(), LogisticRegression(penalty='l1')])
ensemble5.add_meta(LogisticRegression(penalty='l1'))
ensemble5.fit(X_train, y_train)
preds5 = ensemble5.predict(X_test)
print(accuracy_score(preds5, y_test))

C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\Harsh\.LSC\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarn

0.9411764705882353


**Best Accuracy and lowest RMSE**

In [56]:
clf6 = lgb.LGBMClassifier()
clf6.fit(X_train, y_train)
y_pred6 = clf6.predict(X_test)
print('Best Accuracy:', accuracy_score(y_pred6, y_test))
print('Best RMSE:', math.sqrt(mse(y_pred6, y_test)))
print('Model Probability', clf6.predict_proba(X_test))

Best Accuracy: 0.9607843137254902
Best RMSE: 0.19802950859533486
Model Probability [[9.92895575e-01 7.10442470e-03]
 [9.90582337e-01 9.41766315e-03]
 [3.96103733e-04 9.99603896e-01]
 [4.29257886e-02 9.57074211e-01]
 [1.38096043e-02 9.86190396e-01]
 [8.97433300e-01 1.02566700e-01]
 [9.99706149e-01 2.93850849e-04]
 [4.21166166e-01 5.78833834e-01]
 [9.99969548e-01 3.04520400e-05]
 [1.47093433e-03 9.98529066e-01]
 [3.14089924e-02 9.68591008e-01]
 [9.95383026e-01 4.61697429e-03]
 [9.99527336e-01 4.72664399e-04]
 [9.66495923e-01 3.35040767e-02]
 [9.03564025e-01 9.64359753e-02]
 [1.33791400e-02 9.86620860e-01]
 [1.18963502e-01 8.81036498e-01]
 [9.97293239e-01 2.70676133e-03]
 [6.34687661e-01 3.65312339e-01]
 [9.99811945e-01 1.88054975e-04]
 [9.98516299e-01 1.48370135e-03]
 [8.99464109e-01 1.00535891e-01]
 [9.79547032e-01 2.04529675e-02]
 [2.23713155e-04 9.99776287e-01]
 [9.97930108e-01 2.06989167e-03]
 [8.88753727e-01 1.11246273e-01]
 [9.98377140e-01 1.62286002e-03]
 [1.98224846e-04 9.9980177

*Let's try Tuning*

In [52]:
from sklearn.model_selection import RandomizedSearchCV
def hypertuning_rscv(est, p_distr, nbr_iter,X_train, y_train):
    rdmsearch = RandomizedSearchCV(est, param_distributions=p_distr,n_jobs=-1, n_iter=nbr_iter, cv=9)
    rdmsearch.fit(X_train, y_train)
    ht_params = rdmsearch.best_params_
    ht_score = rdmsearch.best_score_
    return ht_params, ht_score

p_dist2 = {'learning_rate':[0.1, 0.2, 0.3, 0.5, 0.7, 0.9, 0.95],'num_leaves':[10, 15, 20, 25, 30, 35, 40, 47, 55, 60, 65, 70, 75, 80]}
lgb_parameters, lgb_ht_score = hypertuning_rscv(lgb.LGBMClassifier(), p_dist2, 40, X_train, y_train)
print(lgb_parameters, lgb_ht_score)

C:\Users\Harsh\.LSC\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


{'num_leaves': 20, 'learning_rate': 0.3} 0.9054726368159204


In [54]:
clf9 = lgb.LGBMClassifier(num_leaves= '20',learning_rate= '0.3')
clf9.fit(X_train, y_train)
y_pred9 = clf9.predict(X_test)
print('Best Accuracy:', accuracy_score(y_pred9, y_test))
print('Best RMSE:', math.sqrt(mse(y_pred9, y_test)))

Best Accuracy: 0.9215686274509803
Best RMSE: 0.28005601680560194


***We have Best Model Performance of Light GBM with 96% Accuracy and with probabilities as indicated above (First one for target 0 and second one for target 1). Please Check In [56]. One can try with manual inputs. The order of the output is unchanged, i.e. output 0 denotes that Team which bowled first wins, as predicted by model***